### DISCUSSION TITLE
Shareservice endpoint does not exist

### DISCUSSION THREAD

**Discussion by LFNPRVKD** [photobooth-app/photobooth-app#312] [Q&A] [answered]
I am having trouble getting the shareservice to work on my VPS. I have all the files; I know that the php and apache2 are working. I am not sure about sqlite3 I know I have installed it correctly but I do not have a way to test it. There is nothing in the photobooth logs about the photos sending to the VPS. however, I have enabled sharing in the photobooth configuration, as well as matching the api key and the URI. the php.log on the VPS shows "endpoint does not exist!" I have granted read write permissions for www-data as well as my own user on the VPS. I have read all the discussions and issues on the github pertaining to this, I am lost now and I don't know what to try next any help is greatly appreciated.

**Comment by mgineer85** [comment]
in dl.php endpoint does not exist is emitted, when you just call the script.
what if you browse to
dl.php?action=info
or
dl.php?action=list
?
Did you set the API key in the header of the dl.php script?

If the above calls work, at that point it is also ensured that sqlite is working and you should see a database in the VPS.

**Reply by LFNPRVKD** [reply]
?action=info Shows 
```
{"version":2,"name":"photobooth-app fileupload extension"}
```

?action=info Shows 
```
Array
(
    [file_identifier] => b454098ef243a191b592d56c7e449a1a
    [filename] => 
    [last_modified] => 2024-07-03 06:23:14
    [status] => pending
)
Array
(
    [file_identifier] => 7b5ba8e483ad8e970a8a640c1c1034af
    [filename] => 
    [last_modified] => 2024-07-03 06:08:46
    [status] => pending
)
```

yes the api key matches

**Reply by mgineer85** [reply]
So that seems fine.
Any errors in the apps log?
Can you upload the logfile here?

**Reply by LFNPRVKD** [reply]
I tried updating the photobooth software when I noticed it was out of date. not change, attached is the logfile
[Logfile.txt](https://github.com/user-attachments/files/16091287/Logfile.txt)

**Reply by mgineer85** [reply]
Did you see the error in the logs? I would look into that. also you need to add th dl.php to the URL.

2024-07-03 00:08:43,633 [    INFO] checking shareservice api endpoint (shareservice.py:31)
2024-07-03 00:08:43,641 [ WARNING] error checking shareservice api endpoint: Invalid URL 'weddingphotos.YYY': No scheme supplied. Perhaps you meant https://weddingphotoset.YYY? (shareservice.py:35)
2024-07-03 00:08:43,641 [   ERROR] shareservice init was not successful. start service aborted. (shareservice.py:63)

**Reply by LFNPRVKD** [reply]
my php-error.log now shows 
```
[04-Jul-2024 19:27:01 UTC] PHP Fatal error:  Uncaught Error: Call to undefined function imagecreate() in /var/www/html/dl.php:40
Stack trace:
#0 /var/www/html/dl.php(268): text_to_image()
#1 {main}
  thrown in /var/www/html/dl.php on line 40
```
I am very sorry I am having so much trouble, truly thank you for all your help so far

**Reply by mgineer85** [reply]
You need the gd libraries enabled on your server in php.

**Reply by LFNPRVKD** [reply]
I just want to confirm that I have corrected the uri issue from earlier 
```
2024-07-04 16:40:57,631 [ WARNING] error checking shareservice api endpoint: Invalid URL 'weddingphotos.ddns.net/dl.php': No scheme supplied. Perhaps you meant https://weddingphotos.ddns.net/dl.php? (shareservice.py:35)
2024-07-04 16:40:57,631 [   ERROR] shareservice init was not successful. start service aborted. (shareservice.py:63)
2024-07-04 16:40:57,632 [    INFO] started ShareService (container.py:86) 
```
I still get an error that the share service was not started, but the next line says it was. is there a test I can do to see why I am still getting this error? or any other logs I can look at to see why I am still getting this?

I receive this from the share service
![image](https://github.com/photobooth-app/photobooth-app/assets/32719806/dd9873b8-5286-496e-a120-ff1b15370133)

**Reply by mgineer85** [reply, marked as answer]
You need to add the http or https before the domain as suggested.

**Reply by LFNPRVKD** [reply]
Thank you so very much! that solved all my problems.

**Comment by WebfinityDesign** [comment]
I'm pulling my hair out on this.

Situation 1: on the local network on my homeserver it is working fine and very fast, almost instant.
Situation  2: when I try to access the same homeserver from outside I get the exact same green runtime error
Situation 3: tried other webhosting. Same green runtime error. 

On situation 2 and 3 dl.php?action=info responds with {"version":2,"name":"photobooth-app fileupload extension"} and an sqlite3 file is created.
In all cases I checked the API-key many times. I'm sure it is matching.

The shareservice.py reports a time=8 in the log. I tried raising that to 25, but the only thing happening is that it reports a timeout=25

What else can I try?

**Reply by mgineer85** [reply]
Could you please also share the complete log from the Webserver and from the PhotoBooth ? Thank you

**Comment by WebfinityDesign** [comment]
[latest.txt](https://github.com/user-attachments/files/16120596/latest.txt)

This is the log from the PhotoBooth. The error I have seen on the server was with a journalctl command. I cannot find it in a file, but I copied the error message. It was: 

error connecting to service: HTTPSConnectionPool(host='xxxxxxxxxxxxx.xxx', port=443): Read timed out. (read timeout=8)

**Reply by mgineer85** [reply]
Is your host somewhere in your network/using dynamic dns services? Then the networks router would probably limit the communication to internal nodes via external domains.

Also there is a lot errors about WLED not setup correctly. It cannot connect to the given device port...

**Comment by WebfinityDesign** [comment]
Are you talking about the internet connection to the photobooth or the internet connection to the home server?
The access from outside to my home server on my home network is running through a Cloudflare tunnel. Could that be an issue?
The photobooth itself is through the router directly connected to the internet as far as I know.
I also tried putting the dl.php on an separate webhoster. On that one dl.php?action=info responds with {"version":2,"name":"photobooth-app fileupload extension"} but when I try to download a photo I get the green runtime error.
Maybe it is because it's a cheap hosting. Where do you host the dl.php yourself?

About the wled errors: I temporarily disconnected the D1 mini, that's why the errors. I disabled it now in the admin section.

UPDATE: to rule out the network I connected the photobooth through a 5G connection and also tried to download the photo through a 5g connection. The result is the same green runtime error. I guess the only thing left would be the webhosting.

UPDATE 2: I disabled IPV6 on the photobooth wifi connection to rule out any conflicts with IPV4/IPV6, but that also makes no difference

**Reply by mgineer85** [reply]
In the end your PhotoBooth host cannot connect to the given endpoint. Can you ping it from the booth? Can you load the url from the photobooth host browser?

**Comment by WebfinityDesign** [comment]
I can ping the URL.
If I try to open dl.php in the browser on the photobooth it responds with a 406 error.
If I open dl.php?action=info in the photobooth browser it responds wirh {"version":2,"name":"photobooth-app fileupload extension"} .

I tried it on my home server and on the webhosting. Both exact the same result.

**Reply by mgineer85** [reply]
I cannot think of a way, why python would not successfully request a url but the browser does. I am sorry, but there must be something off in the configuration or system beyond.

**Comment by WebfinityDesign** [comment]
Any idea how to troubleshoot?

Can I manually test something? Put a photo on a certain location on the photobooth and see if it will download it?
Where is the photo stored that would be ready for download with the link? Maybe I can monitor that location to see if a photo is placed in there for download.

Does the 406 error give any clue, or is it expected because there is no photo to download?

Could it be something in the configuration of the host that is different than it is on your hosting?

This is what I found in the error log of the SQLite3 SB:

(1) QXcbConnection: XCB error: 148 (Unknown), sequence: 228, resource id: 0, major code: 140 (Unknown), minor code: 20 (, :0)
Maybe that was a one-time error?

The tables in the database are present: upload_request, file_identifier, filename, last_modified and status.

![SQLIte3](https://github.com/photobooth-app/photobooth-app/assets/13224944/585a1388-90dc-442d-b695-88becf8ab836)

Is that of any help?

**Reply by mgineer85** [reply]
Well, there are some files in the database uploaded. These should have worked to download.
There is also types of "job_assigned" which means the booth was listening and got the task to upload a file assigned.
But then the booth was not able to process further, the reasons for this are unknown to me based on the logs provided earlier.

I still think it might be related to the network connection.
The booth is creating a long running php request that is used to listen for jobs. It will last for 240seconds maximum.
Maybe cloudflare cuts this?

406 is good, because if there are no actions defined, it will just send 406.

**Reply by mgineer85** [reply]
2024-07-07 17:15:44,742 [ WARNING] error connecting to service: HTTPSConnectionPool(host='xxxx.tvnlbox.nl', port=443): Read timed out. (read timeout=8) (shareservice.py:85)

Many of these errors - there should be read timeouts only in exceptional cases.

**Comment by WebfinityDesign** [comment]
I tried it on a laptop and on a desktop machine, both running Ubuntu 22.04.
On the desktop, if I take a picture with the virtual camera, after a page refresh from the browser, I get :

[   ERROR] Exception in ASGI application

followed by some lines about routing. I hope that can give you a clue.
The logfile is attached.

[latest.txt](https://github.com/user-attachments/files/16149911/latest.txt)

**Reply by mgineer85** [reply]
Hm, it doesn’t seem to be about the apps code but an issuecwith the underlying libs or the combination with python 3.10 on the system.
Did it work on that system earlier? Did you update anything?

**Reply by mgineer85** [reply]
Seems to be related to this:
https://github.com/encode/uvicorn/issues/2271
and was already fixed in https://github.com/encode/uvicorn/pull/2276

So you seem to be on older version for uvicorn?

**Comment by WebfinityDesign** [comment]
This is on the desktop pc (Ubuntu 22.04) I use for testing purpose. It is easier to monitor photobooth app in the terminal.
I never managed to download an image via the QR code on this machine.
Updating uvicorn did solve the ASGI error.

Current status:
I placed the dl.php on a vultr VPS. On this VPS it responds with "succesfully connected top dl.php"
I start the photobooth app from the terminal on my desktop pc.

- phototbooth app is connecting succesfully with dl.php
- in the terminal I can see that the file is being uploaded to the VPS, but it gives a runtime error on shareservice.py line 167 (terminal output attached) In the photobooth error.log you see more or less the same (also attached)
- the image file is being uploaded to the VPS. I can view the image. But it has a (I think) temporary name: phpzdS8LR.jpg (attached)
- in the database is a file_identifier for the job but no filename (screenshot attached)
- in the php-error.log on the server you can see that it is trying to upload the file but it stops with a runtime error (php-error.log attached)

Remarks. I don't know if the next things are an issue:
- the Vultr VPS drops the SFTP-connection  to the terminal after a short period of inactivity (+/- 5 minutes) and then it takes some time to reconnect. (Possibly this is only when using the terminal)
- the last_modified time in SQLite3 on the Vultr server is 2 hours behind my local time, although the timezone and locales on the VPS are correctly set for my region

[latest.txt](https://github.com/user-attachments/files/16160349/latest.txt)
[terminal.txt](https://github.com/user-attachments/files/16160351/terminal.txt)
[php-error.log](https://github.com/user-attachments/files/16160352/php-error.log)
![phpzdS8LR](https://github.com/photobooth-app/photobooth-app/assets/13224944/24404ac5-d5ca-40b9-af52-730c318d4269)
![database](https://github.com/photobooth-app/photobooth-app/assets/13224944/e3d0781f-5ca1-4e8a-a4cf-f1f277201be1)

I hope with this you have enough information to find the bottleneck
